### PR TITLE
fix(local-runtime): unify memory accounting and effective-window MTP

### PR DESCRIPTION
**Pull request by jquesnelle**
## What does this PR do?

Make the managed local runtime use consistent memory accounting and choose MTP batch settings at the context window it will actually launch.

Previously, initial admission and spill classification could omit runtime/projector/logits overhead even though the window-selection path priced it. MTP batches were chosen before restoring a previously grown context, allowing unnecessary CPU spill when lean MTP would fit at the same window. Growth could also report a grant without verifying that the regenerated preset contained it.

## Changes

- Add a shared complete-footprint calculation and launch planner used by catalog selection, catalog rows and generated presets.
- Preserve the window-first policy and integrated MTP. Use larger prefill batches when appropriate; use lean MTP when it preserves a larger window or avoids spill at the same effective window. Reevaluate persisted grown windows before giving up their grant.
- Gate growth on the complete planned footprint and read back the materialized preset after refresh before reporting a new window.
- Price optional external DSpark drafts from their own GGUF metadata, including weights, full target-window f16 context and runtime/logits buffers. Omit the draft if it cannot fit without shrinking the base-model grant.
- Use explicit admitted model paths in preset-only router discovery so a refused staged file cannot autoload with stock settings. Regenerate legacy pathless presets.
- Write presets atomically and retain refusal/spill facts in INI comments. Cover dense spill and spill-to-resident status transitions through the authenticated status endpoint, with the server's context grant remaining authoritative.
- Update the user-facing local-model documentation. No new configuration keys or tuning UI.

## Related work and contributor credit

This consolidates the overhead-accounting contribution in #102993 by @infinitycrew39 and the restored-window MTP contribution in #106897 by @KoNit-K. Both contributors are credited in the commit.

- **#106897 — superseded; closes when this PR merges.** Its MTP re-check is handled by the shared effective-window planner, including initial launch and restart.
- **#102993 — partially incorporated; closes when this PR merges.** This PR includes overhead-aware admission, but deliberately does not adopt the proposed blanket 40% host-RAM reserve. Track the unresolved host-RAM safety work in #102865 for a separately validated approach; closing this implementation proposal does not resolve the underlying issue.
- **Partially addresses #106895, section 3.** In-chat spill notifications and persistent parameter overrides remain separate. The existing Local Models placement pill consumes the corrected persisted facts; this is not live measurement of GPU placement.
- **Related to #102865, not a complete fix.** The reporter-shaped synthetic budget still recommends the spilled 35B model. Host-RAM pressure and Vulkan device selection remain unresolved, and the reported desktop freeze has not been reproduced on native matching hardware.

The closing references below retire the two older implementation proposals when this PR merges into `main`. Issues #102865 and #106895 remain open for the unresolved problems.

Closes #106897
Closes https://github.com/NousResearch/hermes-agent/pull/106897#issuecomment-5630084364
Closes #102993
Closes https://github.com/NousResearch/hermes-agent/pull/102993#issuecomment-5630084597

## Validation

- **205 passed, 0 failed, 1 Linux-only skipped** across 16 affected runtime/catalog/status test files on Windows 11, using `scripts/run_tests.sh` with file retries disabled.
- Ruff on the affected Python files and `git diff --check`: passed.
- Independent review: no remaining reported logic/security defects after optional-draft admission was corrected. The suggested exact draft-memory boundary coverage is included.
- Desktop `npm run build`: passed, with non-blocking bundler code-splitting/plugin-timing warnings.
- Native **RTX 5090 / llama.cpp b10679 / Qwen3.8-27B-UD-Q4_K_M**: exercised stacked and lean MTP at a 65,536-token window; retained the existing 221,184-token capacity-planned window and completed prompts through 180,224 tokens. This is inference evidence, not a claim that planner spill flags measure actual WDDM residency.
- Isolated real managed-supervisor run: **65,536 -> 73,728 -> restart at 73,728**, integrated MTP retained, only the admitted model visible, and a real Hermes write-file/read-file task verified by file read-back. The planning budget in that run was synthetic 24 GiB on the physical 5090, not native 24 GiB certification.
- Manual desktop test after `hermes desktop --force-build`: loaded the model and completed a multi-turn structured-tool conversation. No inference failure was recorded; the final review turn was interrupted. Model instruction-following quality is outside this runtime fix.

### Repeat the affected suite

```bash
bash scripts/run_tests.sh -j 3 \
  tests/hermes_cli/test_context_policy.py \
  tests/hermes_cli/test_catalog_variants.py \
  tests/hermes_cli/test_local_recommendation.py \
  tests/hermes_cli/test_local_growth.py \
  tests/hermes_cli/test_boot_preset_staleness.py \
  tests/hermes_cli/test_local_models_routes.py \
  tests/hermes_cli/test_local_quickstart.py \
  tests/hermes_cli/test_local_runtime.py \
  tests/hermes_cli/test_local_runtime_picker_row.py \
  tests/hermes_cli/test_local_runtime_updates.py \
  tests/hermes_cli/test_budget_source.py \
  tests/hermes_cli/test_unified_pool_quirk.py \
  tests/hermes_cli/test_local_preset_admission.py \
  tests/hermes_cli/test_runtime_machine_scope.py \
  tests/hermes_cli/test_hf_browse.py \
  tests/hermes_cli/test_catalog_json.py -q
```

## Scope and remaining validation

The Q4 quality floor, desktop VRAM reserve, embedded sampling defaults and context-growth policy are retained. This PR does not add a flat context cap, change host-RAM reserve policy, expand the catalog, repair downloads or add advanced controls.

Native 16/24 GiB cards, the separate 48 GB machine, Linux Vulkan and Metal have not been validated by these local runs. Unified-memory cases have policy-test coverage only. The full repository suite was not run locally; CI status is reported separately.

The manual/native tests used checkout base `72a3277cd7937fd0f0a2a3e3fddbed21d7b1c8bd`. Before publication, the touched existing files were compared with fetched main `05d705dd69` and had no intervening changes. No unrelated commits are included in the PR.

### CODE REVIEW

**Review by andrexibiza — COMMENTED** [review]
Reviewed exact head `0316d3d404a6dbb3b03e27f4f4728b673cc3de11` end-to-end against live `main@bffa5f75e45f41045c000c9516f08cada3d56eb4`, including the planner/catalog/preset/growth/supervisor/status seams, the affected tests, current hosted checks, and the overlapping local-runtime work.

## Verification

I do **not** see a code-level blocker in the memory/MTP consolidation itself.

- The accounting is now one shape instead of several almost-the-same shapes: `footprint_bytes()` is threaded through admission, `initial_window()`, effective-window restore, catalog choice, and growth instead of letting runtime/projector/logits overhead disappear at one of those boundaries.
- MTP posture is decided at the window that will actually launch. I traced the initial and restored-window cases, including the both-spill control inherited from the #106897 defect class; the planner does not silently shrink a previously granted window to save stacked prefill.
- Growth no longer treats “planner said grow” as settlement. It regenerates the preset and reads the materialized decision back before reporting the grant, which closes the plan-vs-realization gap.
- Refused staged models stay policy-owned: the preset records the refusal, the supervisor does not also hand the router `--models-dir` when a preset exists, and explicit model paths prevent a refused file from reappearing through stock-fit autoload.
- I traced the optional external DSpark draft path specifically. The admission check prices the draft from its own GGUF metadata, full target-window f16 KV, runtime overhead, and logits; combined host+device capacity is checked, while target KV/runtime + the optional draft must fit device memory. Target weights remain the spillable axis. The exact boundary/negative-control coverage matches that contract, so I do not see a second hidden “draft gets free memory” accounting hole here.
- Atomic preset replacement plus persisted refusal/spill facts gives the Desktop/status surface a durable record of the launch policy instead of reconstructing it from filenames or `override-tensor` alone.

Hosted evidence is unusually clean for this train: exact head CI `34566911120`, Docker `34566910686`, and Nix `34566910663` are all green. This PR has one surviving commit, so the **current object is 1/1 hosted-green**.

## Merge gate: current-main integration is not proven yet

The one thing I would not wave through is topology/evidence, not the implementation above. The branch commit is rooted at merge base `72a3277cd7937fd0f0a2a3e3fddbed21d7b1c8bd`; against live `main@bffa5f75e45f41045c000c9516f08cada3d56eb4` it is **1 commit ahead / 554 commits behind**. I independently checked the directly touched local-runtime production files for main-side drift and did not find an overlapping edit in that interval, which is why GitHub can still report this mergeable. But exact-head green on the old branch tree is not evidence for the current-main integration tree in a repository moving this quickly.

Required before merge: bring current `main` into this branch **without discarding/recreating the existing contribution or its co-author trailers**, then require CI + Docker + Nix to execute and pass on the resulting exact head. If that integration is content-neutral for this diff and those checks stay green, I consider the implementation itself clear.

## Interlocks / credit

The consolidation story is coherent and the commit preserves the right provenance: `jquesnelle` owns this carrier, with `infinitycrew39` and `KoNit-K` retained as co-authors.

- **#106897 (`KoNit-K`)** is genuinely superseded by the shared effective-window planner here. Close it only as replacement/supersession, not as disposable duplicate work.
- **#102993 (`infinitycrew39`)** is only partially incorporated: the overhead-aware admission work is carried; the proposed 40% host-RAM reserve is deliberately not. That residual defect class must remain owned by **#102865** if the implementation PR is closed.
- **#106895** is only partially addressed (§3). Spill disclosure and persistent user parameter/override surfaces remain separate work and should stay open.
- **#102865** remains the host-pressure / device-selection problem record; this PR does not claim to settle it.

That separation matters: this carrier fixes accounting consistency and effective-window realization; it should not accidentally close the still-unsettled host-RAM or UX sides of the shape.

Substantial work here. Once the current-main object is actually proven, this is a strong consolidation rather than another parallel local-runtime policy path. 🚀

**Comment by 3x3xX3N0N** [comment]
#102993 was closed as partially incorporated when this merged. Two unified-pool parts of it don't appear to have carried over, and I wanted to check whether that's deliberate rather than let them look handled from the outside.

On current `main`:

- `hardware.py:35` still has `_UMA_HEADROOM_FRACTION = 0.20`, and `_uma_budget` (line 254) is still `usable = max(0, int(base * (1 - _UMA_HEADROOM_FRACTION)))`.
- Outside planning mode `probe_budget` passes `ram_avail` as `base` (line 295), so the reserve held back for the OS is 20% *of whatever is free right now*. As the machine fills, the amount reserved for the OS shrinks, and the same Mac reports a different budget minute to minute. That is the opposite of what a reserve should do.
- A flat fraction is also awkward at both ends: 20% of a 256 GiB Mac Studio withholds 51 GiB from an OS that will not use a fraction of it, while small machines are squeezed.
- I don't see a UMA OS free-memory floor in `estimator.py`.

@infinitycrew39's #102993 proposed a clamp for the first and a floor for the second. I'm not proposing to re-land their patch — it's their work and their call. I'm flagging that on a unified pool these are still live.

Separately, and not covered by either PR: nothing in the tree reads `iogpu.wired_limit_mb`, the macOS ceiling on how much unified memory the GPU may actually wire. `0` there means "use the system default", commonly documented as ~75% of RAM, and Metal reports the effective figure as `recommendedMaxWorkingSetSize` (llama.cpp prints it at `ggml_metal_init`). A budget above that ceiling can be admitted and then fail to wire or fall into swap, so an operator who has *lowered* the ceiling is currently planned past. Happy to open a separate issue rather than widen this thread.

Context: I authored #105312, which touches the same file but only the total read, and does not overlap the headroom policy.
